### ISSUE TITLE
CMPLOG feature for GCC

### ISSUE THREAD

**Issue by Zherya** [AFLplusplus/AFLplusplus#744]
Hello! I'm interested in CMPLOG for GCC, are there any plans or developing process of such feature for GCC?

If no, perhaps, I'm interested in developing it in my own.

**Comment by vanhauser-thc** [comment]
I asked @ButcherAdaCore if they wanted to do it some weeks ago, not sure what their plans are, if they have already done it, or if it is currently a  "wont do" :) Lets see what he says.
CMPLOG for the gcc_plugin would be really good and important.

**Comment by ButcherAdaCore** [comment]
Hi both - it is very much on our todo list. 

I'll chase it today and see where we are with it.

**Comment by ButcherAdaCore** [comment]
@vanhauser-thc @Zherya - we're currently clearing a few higher priority tasks and then CMPLOG for GCC will be next. We're predicting a reasonably quick turn around (less than 1 week to implement, test and upstream).

Unfortunately I can't give you a more concrete estimate at this stage but will let you know as soon as work is underway.

No problem at all if in the meantime somebody else wants to commit to the task.

**Comment by vanhauser-thc** [comment]
@Zherya if you want to have a go at it, here is a description I gave to Paul:
```
it is really easy to instrument.

for a cmp/fcmp on 8 bit, both arguments are instrumented to
__cmplog_ins_hook1(arg0, arg1, attr);
where attr is <, >, <=, >=, == or !=
1 bit = equal
2 bit = greater
3 bit = lesser
4 bit = Floating Point comparison
 so ... 5 = integer lesser equal comparison, 11 = floating point lesser comparison

and for 16 bit it is __cmplog_ins_hook2, 32 bit __cmplog_ins_hook4, 64 bit __cmplog_ins_hook8
and anything bigger than 64 bit (if gcc plugins supports that)
are casted to u128 and sent to __cmplog_ins_hookN which then has a 4th
parameter, which is the number of (bytes - 1), so a

  fcmp u128, u128
  jle xxx

would be
__cmplog_ins_hookN(arg0, arg1, 1 + 4 + 8, 15)

then finally there is one more instrumentation:
if a function call has at least 2 parameters and the first two
parameters are both pointers then both parameters are given to
__cmplog_rtn_hook(ptr0, ptr1), other parameters ignored.

in llvm basically 30 minutes work 

To activate I would propose the AFL_GCC_CMPLOG environment variable.
```

**Comment by Zherya** [comment]
@vanhauser-thc @ButcherAdaCore Thanks for the explanations! I suppose, if @ButcherAdaCore team plans to develop CMPLOG so soon, it is better to leave it to them. Because, I believe, they know GCC much better than me:)

Perhaps, I may choose some another absent gcc_plugin feature for developing it slowly:)

**Comment by domenukk** [comment]
Would it make sense to use an additional, unified `AFL_CMPLOG` variable that just enables it whenever available? Could make switching modes easier for the user

**Comment by vanhauser-thc** [comment]
@Zherya if you need help just contact us (e.g. by opening a new issue - but ButcherAdaCore with an @ reference as we ourselves do not have much gcc_plugin knowledge)
@domenukk maybe, maybe not? dunno
@ButcherAdaCore looking forward to gcc cmplog!! :)

closing this for now :)

**Comment by ButcherAdaCore** [comment]
@vanhauser-thc @Zherya just a quick note to state that we have started this work. 

The first pass (calls hooks at compares) is currently being tested... we will then look at the other aspects.

**Comment by vanhauser-thc** [comment]
cool!

btw for testing please use current dev with afl-fuzz options `-Z -l 3` - for effectiveness reasons not every queue entry is used for cmplog. -l 3 forces that though so for testing this is a good thing to set.

**Comment by vanhauser-thc** [comment]
@ButcherAdaCore what is the status? :)

**Comment by ButcherAdaCore** [comment]
Hi @vanhauser-thc - to date we have implemented the pass that calls hooks at compares. The one that calls hooks at function calls taking pointers is yet to be implemented. This was actually done a couple of weeks ago, however, due to other priorities we have only just started testing the additional compilation pass (and benchmarking the capability). I expect we'll have our testing and benchmark results completed by the end of the week.

@domenukk - in answer to your comment: I agree it makes sense to have a unified environment switch to enable the additional pass - `AFL_CMPLOG` would seem like a good candidate. Unless anyone objects we'll implement it this way.

We're aiming to complete our testing by the end of the week and will likely need a few more days to complete the instrumentation around pointers. Hopefully we can get this done within the next few weeks, however, that depends on if higher priority tasks come in... :-)

**Comment by vanhauser-thc** [comment]
@ButcherAdaCore please use `AFL_GCC_CMPLOG` - this otherwise breaks the format. but it is fine if you check additionally for `AFL_CMPLOG`

**Comment by ButcherAdaCore** [comment]
@vanhauser-thc no problem `AFL_GCC_CMPLOG` it is! :-)

**Comment by vanhauser-thc** [comment]
@ButcherAdaCore last ping was over 3 months ago :) how far along are you with the implementation? :)

**Comment by ButcherAdaCore** [comment]
Morning @vanhauser-thc, sorry for not providing an update on this sooner.

Unfortunately we've made little progress with this since my last update. Our compiler team is fully occupied working on another higher priority project and the CMPLOG work has had to take a back seat. As previously stated we did implement and test the first compiler pass, however, we found problems with the code and didn't progress as far as fixing them.

At this stage we're not expecting to free up the required resources until next year. We can submit the work we have done so far on a branch for somebody else to pick up, however, it would be unlikely that we would find ourselves in a position to provide any support with the development in the near future - sorry, we're just currently snowed under.

Let me know if this is of interest to you?
